### ISSUE TITLE
Union types as props for JSX/TSX components behave strange

### ISSUE THREAD

**Issue by donaldpipowitch** [Microsoft/TypeScript#15002]
**TypeScript Version:**  2.2.2

**Code**

![image](https://cloud.githubusercontent.com/assets/1152805/24651132/f39a279e-192c-11e7-95b7-717632c82285.png)

```ts
interface Foo {
  type: 'foo';
  hello: string;
}

interface Bar {
  type: 'bar';
  world: string;
}

type FooOrBar = Foo | Bar;

const FooOrBarComp: SFC<FooOrBar> = props => <div />;


const test1: FooOrBar = { type: 'foo' };          // property 'hello' is missing - correct
const Test1 = () => <FooOrBarComp type="foo" />   // property 'hello' is NOT missing - wrong

const test2: FooOrBar = { type: 'bar' };          // property 'world' is missing - correct
const Test2 = () => <FooOrBarComp type="bar" />   // property 'world' is NOT missing - wrong

const test3: FooOrBar = { type: 'foo', hello: '123' };        // works - correct
const Test3 = () => <FooOrBarComp type="foo" hello="123" />   // property 'hello' does not exist - wrong
```

**Expected behavior:**

`Test1`, `Test2` and `Test3` should behave like `test1`, `test2` and `test3`.

**Actual behavior:**

`Test1` and `Test2` are false positives. No error is shown, but the property `world` is missing.

`Test3` is false negative. An error is shown, but the properties are correct.

**Comment by cevek** [comment]
In nightly TS 2.3 it works as intended:
```ts
error TS2322: Type '{ type: "foo"; }' is not assignable to type 'FooOrBar'.
  Type '{ type: "foo"; }' is not assignable to type 'Foo'.
    Property 'hello' is missing in type '{ type: "foo"; }'.
```
